### ISSUE TITLE
New sampling mechanisms (Sobol, Giustino)

### ISSUE THREAD

**Issue by mesonepigreco** [SSCHAcode/python-sscha#82]

**Comment by diegomartinez2** [comment]
Geting ready also the new branch "Diegom_sobol" for python_sscha for calling the Sobol sampling from CellConstructor "Diegom_sobol" branch. Also trying to maintain retro-compatibility, so the sobol flag is False by default.

**Comment by diegomartinez2** [comment]
Testing "Diegom_sobol" branches (for python-ssha and CellConstructor). It doesn't look like the Sobol sampling improves the calculations. As I implemented it, the Sobol function always gives the same sequence. Maybe it will need improvements or some random "salt".

**Comment by mesonepigreco** [comment]
Hi Diego,
Thanks for your efforts.
Indeed, it is quite difficult to estimate how much this can improve. I would say we can observe several aspects:
 1) In a long minimization, how many populations are required to converge with the same number of configurations?
 2) When we are close to convergence, what is the minimum number of configurations required to have the gradient to go to zero without the ensemble going outside the statistical criterion?

If you can try to plot somethibg like these two values it would be important to understand if sobol improves, does nothing or worsen the convergence.

I would use the SnTe force field, as it is superfast to compute to make these tests, but if you have better ideas, please, go on!

**Comment by diegomartinez2** [comment]
Found a error in the way I introduced the Sobol sequence. I was following the random generation, but in SSCHA the random generator is later divided between the modes. This is not a problem with random numbers but is a problem with Sobol sequence. I'm changing the code to create one Sobol sequence per mode. So in the end is one Sobol sequence per mode per configuration.

**Comment by diegomartinez2** [comment]
There is another issue with Sobol. In this case is about transforming the Sobol sequence into a gaussian shape. In the process of transforming the sequence into gaussian, only the points inside the unit radius circle are taken so there are points of the Sobol sequence that are eliminated. This is not a problem due to the fact that an uniform set of points in a square are still uniform if we cut the corners, but the number of points at the end will be less and the idea of making the points powers of 2 gets more complicated. In other words we can generate 2**n points and cut the corners but at the end there will be less points and in extreme cases of low numbers there may be almost nothing at the end.

**Comment by mesonepigreco** [comment]
No, this way is super inefficient:

Simply use the standard algorithm for gaussian numbers:
```
# u1 and u2 are uniform in [0,1)
r = -np.sqrt(2 * np.log(u1))
theta = 2*np.pi * u2
random1 = r * np.cos(theta)
random2 = r*np.sin(theta)
```

So you get 2 random numbers for each pair of Sobol numbers. Then it is ok since we will generate powers of 2, the number of Sobol numbers are even and you do not discard anything.

**Comment by diegomartinez2** [comment]
The first Sobol number is zero, so the log is not defined. The new method requires half the Sobol numbers and looks different. Still it may be better than random generation for small configurations.

![image](https://user-images.githubusercontent.com/101260978/171347990-5732a322-052d-4433-9c0d-148e592f4985.png)

**Comment by diegomartinez2** [comment]
Yes I see the problem with the direct inversion and also with the Beasley-Springer-Moro method.
![image](https://user-images.githubusercontent.com/101260978/171616967-8c8308b1-8104-4f8a-90c9-393edf0887f7.png)
